java/ql/src/Security/CWE/CWE-524/SensitiveKeyboardCache.qhelp

```diff
@@ -0,0 +1,36 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>When a user enters information to a text input field on an Android application, then by default their input is saved to a keyboard cache, 
+which provides autocomplete suggestions and predictions. If the input field is expected to contain sensitive information, such as a password or banking details, 
+this sensitive data may be leaked to other applications via the keyboard cache.</p>
+
+</overview>
+<recommendation>
+
+<p>For input fields expected to accept sensitive information, an input type such as <code>"textNoSuggestions"</code> (or <code>"textPassword"</code> for a password) 
```

Comment:
Out of curiosity, is there any guarantee that `textNoSuggestions` will for sure not cache or process the input in any way? The documentation for the corresponding [`TYPE_TEXT_FLAG_NO_SUGGESTIONS`](https://developer.android.com/reference/android/text/InputType#TYPE_TEXT_FLAG_NO_SUGGESTIONS) says (emphasis mine):
>  Note the contrast with `TYPE_TEXT_FLAG_AUTO_CORRECT` and `TYPE_TEXT_FLAG_AUTO_COMPLETE`: `TYPE_TEXT_FLAG_NO_SUGGESTIONS` means the IME does not need to show an interface to display suggestions. **Most IMEs** will also take this to mean they do not need to try to auto-correct what the user is typing.

This sounds as if some IMEs might process the input in some way.

Is `textVisiblePassword` maybe better fitting?

Reply:
The source for this is https://github.com/OWASP/owasp-mastg/blob/8cb9ed4a5db21acd40d87212042d98ef7042fbf0/Document/0x05d-Testing-Data-Storage.md#static-analysis-4, which we expect to be accurate in principle.

Also, IMEs are assumed to honor these attributes, but in the end it's up to [the actual implementation](https://developer.android.com/develop/ui/views/touch-and-input/creating-input-method#GeneralDesign). I would expect at least the Android IME to properly not cache the input when suggestions are disabled.